recipe/patches/vopenssl3-pypy3.7.patch

```diff
@@ -0,0 +1,72 @@
+ PY_SSL_DEFAULT_CIPHERS = sysconfig.get_config_var('PY_SSL_DEFAULT_CIPHERS')
+ 
+ PROTOCOL_TO_TLS_VERSION = {}
+@@ -141,6 +142,7 @@
+ OP_SINGLE_ECDH_USE = getattr(ssl, "OP_SINGLE_ECDH_USE", 0)
+ OP_CIPHER_SERVER_PREFERENCE = getattr(ssl, "OP_CIPHER_SERVER_PREFERENCE", 0)
+ OP_ENABLE_MIDDLEBOX_COMPAT = getattr(ssl, "OP_ENABLE_MIDDLEBOX_COMPAT", 0)
++OP_IGNORE_UNEXPECTED_EOF = getattr(ssl, "OP_IGNORE_UNEXPECTED_EOF", 0)
+ 
+ # Ubuntu has patched OpenSSL and changed behavior of security level 2
+ # see https://bugs.python.org/issue41561#msg389003
+@@ -206,6 +208,10 @@
+     if not getattr(ssl, f'HAS_{version.name}'):
+         return False
+ 
++    if IS_OPENSSL_3_0_0 and version < ssl.TLSVersion.TLSv1_2:
++        # bpo43791: 3.0.0-alpha14 fails with TLSV1_ALERT_INTERNAL_ERROR
++        return False
++
+     # check runtime and dynamic crypto policy settings. A TLS version may
+     # be compiled in but disabled by a policy or config option.
+     ctx = ssl.SSLContext()
+@@ -1179,7 +1185,8 @@
+         # SSLContext also enables these by default
+         default |= (OP_NO_COMPRESSION | OP_CIPHER_SERVER_PREFERENCE |
+                     OP_SINGLE_DH_USE | OP_SINGLE_ECDH_USE |
+-                    OP_ENABLE_MIDDLEBOX_COMPAT)
++                    OP_ENABLE_MIDDLEBOX_COMPAT |
++                    OP_IGNORE_UNEXPECTED_EOF)
+         self.assertEqual(default, ctx.options)
+         ctx.options |= ssl.OP_NO_TLSv1
+         self.assertEqual(default | ssl.OP_NO_TLSv1, ctx.options)
+@@ -2281,6 +2288,8 @@
+             rc = s.connect_ex((REMOTE_HOST, 443))
+             if rc == 0:
+                 self.skipTest("REMOTE_HOST responded too quickly")
++            elif rc == errno.ENETUNREACH:
++                self.skipTest("Network unreachable.")
+             self.assertIn(rc, (errno.EAGAIN, errno.EWOULDBLOCK))
+ 
+     @unittest.skipUnless(support.IPV6_ENABLED, 'Needs IPv6')
+@@ -2501,6 +2510,14 @@
+                             sys.stdout.write(err.args[1])
+                         # test_pha_required_nocert is expecting this exception
+                         raise ssl.SSLError('tlsv13 alert certificate required')
++                    if 'UNEXPECTED_EOF_WHILE_READING' == err.reason:
++                        # PyPy OpenSSL3 needs this, on CPython a
++                        # BrokenPipeError is raised which is caught as an
++                        # OSError. In this case do not stop the server.
++                        if self.server.chatty:
++                            handle_error("Test server failure:\n")
++                        self.close()
++                        self.running = False
+                 except OSError:
+                     if self.server.chatty:
+                         handle_error("Test server failure:\n")
+@@ -4613,7 +4630,6 @@
+     if support.verbose:
+         import warnings
+         plats = {
+-            'Linux': platform.linux_distribution,
```

Comment:
Why?

Reply:
It failed for me when i tried it locally